# Στοχαστικές Διεργασίες και Βελτιστοποίηση στη Μηχανική Μάθηση

## 7o Εργαστήριο - *Q-Learning*

- Ονομ/νυμο: Χρήστος Νίκου
- AM: 03400146
- Ιδιότητα: Μεταπτυχιακός φοιτητής Επιστήμης Δεδομένων και Μηχανικής Μάθησης (ΕΔΕΜΜ)
- Ηλεκτρονική Διεύθυνση: christosnikou@mail.ntua.gr / chrisnick92@gmail.com

<h1>Q-Learning</h1>

<p>Στην συγκεκριμένη άσκηση θα μελετήσετε το στοχαστικό αλγόριθμο Q-Learning, χρησιμοποιώντας το έτοιμο πρόγραμμα που σας δίνεται.</p> <p> Στα πλαίσια του παραδείγματος θα εξετάσετε μία υλοποίηση με Q-learning σχετικά με το σύστημα αυτόματης οδήγησης ενός ταξί. Στα πλαίσια του προβλήματος αυτού θα πρέπει να ικανοποιούνται τα εξής:</p>
<ul>
<li>Το ταξί θα πρέπει να αφήνει τον πελάτη στη σωστή θέση</li>
<li>Το ταξί να ακολουθεί τη συντομότερη δυνατή διαδρομή</li>
<li>Να τηρούνται οι κανόνες κυκλοφορίας και ασφάλειας των επιβατών</li>
</ul>

<p>Στα πλαίσια του προβλήματος έχουμε τα εξής χαρακτηριστικά για την επιβράβευση, τις καταστάσεις και τις ενέργειες.</p> 

<h4>Επιβράβευση</h4>
<ul>
<li>Θα έχουμε τη μέγιστη επιβράβευση όταν το ταξί αφήνει έναν πελάτη στην επιθυμητή θέση</li>
<li>Θα υπάρχει ποινή στην περίπτωση όπου το ταξί αφήσει τον πελάτη σε κάποιο λανθασμένο σημείο</li>
<li>Ο agent θα παίρνει μία μικρή σχετικά ποινή στην περίπτωση όπου αργεί να φτάσει στον τελικό προορισμό</li>
</ul>

<p>Γενικά οι παραπάνω αρχές συνοψίζονται στα εξής: "Λαμβάνουμε +20 πόντους για μια επιτυχημένη πτώση και χάνουμε 1 πόντο για κάθε χρονικό βήμα που παίρνει. Υπάρχει επίσης ποινή 10 πόντων για παράνομες ενέργειες παραλαβής και αποχώρησης."</p>

<h4>Πλήθος Καταστάσεων</h4>
<img src="https://storage.googleapis.com/lds-media/images/Reinforcement_Learning_Taxi_Env.width-1200.png">
<p>Στο παράδειγμά μας έχουμε ένα μικρό χώρο 5*5. Από εκεί και πέρα έχουμε 4 προορισμούς και 5 πιθανές θέσεις του πελάτη (παίρνουμε και την περίπτωση ο πελάτης να είναι ήδη μέσα στο τάξι).</p>
<p>Με βάση τα παραπάνω έχουμε 5*5*5*4=500 πιθανές καταστάσεις.</p>

<h4>Πλήθος Ενεργειών</h4>
<p>Έχουμε έξι ενεργειες για το ταξί, οι οποίες είναι οι εξής:</p>
<ul>
<li>0=Νότια</li>
<li>1=Βόρεια</li>
<li>2=Ανατολικά</li>
<li>3=Δυτικά</li>
<li>4=Επιβίβαση</li>
<li>5=Αποβίβαση</li>
</ul>





><b>1.</b> *Να περιγράψετε σύντομα τον αλγόριθμο Q-Learning. Σε ποια προβλήματα θεωρείτε ότι ταιριάζει ως τρόπος εκμάθησης η Ενισχυτική Μάθηση (Reinforcement Learning); Ποια είναι η βασική διαφορά του αλγορίθμου Q-Learning από τους αλγορίθμους Policy Iteration και Value Iteration;*

Ο αλγόριθμος *Q-Learning* είναι ένας *model-free* αλγόριθμος ο οποίος προσπαθεί να βρει τη βέλτιστη πολιτική σε ένα πρόβλημα ενισχυτικής μάθησης <b>*χωρίς να έχει επίγνωση των πιθανοτήτων μετάβασης και των συναρτήσεων αποτίμησης του περιβάλλοντος στο οποίο εφαρμόζεται*</b>. Αυτή είναι και η <b>*βασική διαφορά*</b> του αλγορίθμου Q-Learning από τους αλγορίθμους Policy και Value Iteration. Δηλαδή, το ότι οι δύο τελευταίοι προϋποθέτουν τη γνώση των πιθανοτήτων μετάβασης καθώς και της συνάρτησης αποτίμησης. Οι αλγόριθμοι Policy και Value Iteration μελετήθηκαν στο πρώτο μέρος του 7ου εργαστηρίου. Σε αυτό το σημείο θα δώσουμε μια συντομή περιγραφή του αλγορίθμου *Q-Learning*.

#### Ο Αλγόριθμος *Q-Learning*

Σκοπός του αλγορίθμου *Q-Learning* είναι να προσδιορίσει τη συνάρτηση $Q^*(s,\alpha)$ η οποία δίνει τη συνολική αναμενόμενη αποτίμηση λαμβάνοντας υπ'όψιν τον παράγοντας έκπτωσης $\gamma$ (cumulative discounted reward) στη περίπτωση που ο Agent στην κατάσταση $s\in S$ ενεργεί κατά την $\alpha\in A^1$. Κατά την εκκίνηση του αλγορίθμου ορίζεται ένας πίνακας $Q(s,\alpha)$ με $s\in S,\,\alpha \in A$ ο οποίος αρχικοποιείται με μηδενικές τιμές. Ο πίνακας $Q(s,\alpha)$ εκφράζει την εκτίμηση του αλγορίθμου για την βέλτιστη τιμή $Q^*(s,\alpha)$. Η συνάρτηση $Q(s_t,\alpha_t)$ για μια δεδομένη χρονική στιγμή $t$ δίνεται μέσω της<br><br>

$$Q^{\pi}(s_t,\alpha_t)=\mathbb{E}\biggl(R_{t+1}+\gamma R_{t+2}+\gamma^2R_{t+3}+\dots\,|\,s_t,\alpha_t\biggr).$$

<br><br> Η διαδικασία εκμάθησης και ανανέωσης της συνάρτησης $Q(s,\alpha)$ με σκοπό την προσέγγιση της $Q^*(s,\alpha)$ είναι η εξής:

1. Αρχικά ορίζεται ο πίνακας $Q(s,\alpha)$ ο οποίος είναι διαστάσεων $|S|\times |A|$ και αρχικοποιείται με μηδενικές τιμές.

2. Ξεκινάει ο Agent από μια αρχική κατάσταση $s_0\in S$ και ενεργεί με βάση το σύνολο $A$. Δηλαδή, επιλέγεται ένα $\alpha\in A$ και πραγματοποιείται μια κίνηση.

3. Η ανανέωση της συνάρτησης $Q(s,\alpha)$ γίνεται μέσω της σχέσης<br><br>

$$\tag{1}Q_{new}(s,\alpha) = \underbrace{Q_{current}(s,\alpha)}_{\text{current Q values}}+\underbrace{\alpha}_{\text{learning rate}}\cdot \biggl(\underbrace{R(s,\alpha)}_{\text{Reward}}+\underbrace{\gamma}_{Discount Rate}\cdot\max_{\alpha \in A} Q(s',\alpha')-Q_{\text{current}}(s,\alpha)\biggr)$$<br><br>

Όπως βλέπουμε και απο τη σχέση $(1)$ δεν υπεισέρχεται κάπου η πληροφορία για τις πιθανότητες μετάβασης και τη συνάρτηση αποτίμησης $V$. Το μόνο που λαμβάνει ως πληροφορία ο αλγόριθμος είναι από το περιβάλλον είναι η συνάρτηση αποτίμησης $R:S\times A\to \mathbb{R}$.<br><br>

$^1$ Με $S$ συμβολίζουμε το σύνολο των καταστάσεων και με $A$ το σύνολο των ενεργειών.


#### Λίγα λόγια για την ενισχυτική μάθηση

Στην ενισχυτική μάθηση εντάσσονται τα προβλήματα που λύνονται λαμβάνοντας αποφάσεις και εξετάζοντας την επίδραση που έχουν οι αποφάσεις αυτές στο περιβάλλον που ορίζεται το εκάστοτε πρόβλημα. Το περιβάλλον περιγράφεται από ένα σύνολο καταστάσεων $S$ και ένα σύνολο ενεργειών $A$. Επίσης, δίνεται και μια συνάρτηση $R:S\times A\to \mathbb{R}$ η οποία αντιστοιχεί στην αμοιβή που επιφέρει μια ενέργεια $\alpha \in A$ όταν βρισκόμαστε στην κατάσταση $s\in S$. Ο σκοπός της ενισχυτικής μάθησης είναι να εκπαιδεύση έναν πράκτορα (agent) έτσι ώστε αυτός ξεκινώντας από μια κατάσταση $s\in S$ να λαμβάνει τις ενέργειες $\alpha \in A$ οι οποίες να μεγιστοποιούν τη συνάρτηση αποτίμησης (λαμβάνοντας υπ'όψιν έναν παράγοντα έκπτωσης $\gamma$).<br><br>

Τα προβλήματα στα οποία βρίσκει εφαρμογή η ενισχυτική μάθηση είναι η αυτόματη οδήγηση, αυτόματη μετάφραση και στα παιχνίδια εικονική πραγματικότητας.

<p>Στη συνέχεια θα πρέπει να φορτώσετε τη βιβλιοθήκη gym καθώς και το σχετικό dataset<p>

In [11]:
!pip install cmake 'gym[atari]' scipy

In [12]:
import gym

env = gym.make("Taxi-v3").env

env.render()

+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+



In [13]:
env.reset() # reset environment to a new, random state
env.render()

print("Action Space {}".format(env.action_space))
print("State Space {}".format(env.observation_space))


+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+

Action Space Discrete(6)
State Space Discrete(500)


<p>Παρακάτω ορίζουμε τις συνεταγμένες του ταξί, τη θέση του πελάτη και το σημείο προορισμού</p>

In [14]:
state = env.encode(3, 1, 2, 0) # (taxi row, taxi column, passenger index, destination index)
print("State:", state)

env.s = state
env.render()

State: 328
+---------+
|R: | : :G|
| : | : : |
| : : : : |
| | : | : |
|Y| : |B: |
+---------+



<p>Παρακάτω είναι η μήτρα επιβράβευσης για το state που ορίσαμε στο προηγούμενο βήμα</p>

In [15]:
env.P[328]

{0: [(1.0, 428, -1, False)],
 1: [(1.0, 228, -1, False)],
 2: [(1.0, 348, -1, False)],
 3: [(1.0, 328, -1, False)],
 4: [(1.0, 328, -10, False)],
 5: [(1.0, 328, -10, False)]}

<p> Τρέχουμε το παράδειγμά μας χωρις τη χρήση Q-Learning.</p>

<b><p>2. Τα αποτελέσματα είναι ικανοποιητικά; Πως θα μας εξυπηρετούσε η χρήση του Q-Learning;</p></b>

In [16]:
env.s = 328  # set environment to illustration's state

epochs = 0
penalties, reward = 0, 0

frames = [] # for animation

done = False

while not done:
    action = env.action_space.sample()
    state, reward, done, info = env.step(action)

    if reward == -10:
        penalties += 1
    
    # Put each rendered frame into dict for animation
    frames.append({
        'frame': env.render(mode='ansi'),
        'state': state,
        'action': action,
        'reward': reward
        }
    )

    epochs += 1
    
    
print("Timesteps taken: {}".format(epochs))
print("Penalties incurred: {}".format(penalties))

Timesteps taken: 936
Penalties incurred: 277


><b>*2. Τα αποτελέσματα είναι ικανοποιητικά; Πως θα μας εξυπηρετούσε η χρήση του Q-Learning;*</b>

Όπως βλέπουμε σε μια εκτέλεση του αλγορίθμου που επιλέγει κατά τυχαίο τρόπο μια κίνηση χρειάστηκαν 936 βήματα μέχρις ότου ο πράκτορας να φτάσει στην επιθυμητή κατάσταση και επιβλήθηκαν 277 ποινές. Η χρήση του αλγορίθμου Q-Learning στο συγκεκριμένο πρόβλημα θα μας βοηθούσε έτσι ώστε ο πράκτορας να εκπαιδεύεται κατά τη διάρκεια του αλγορίθμου με αποτέλεσμα να κάνει λιγότερα λάθη.

<p>Τώρα θα προσπαθήσουμε να λύσουμε το πρόβλημά μας με τη χρήση του Q-Learning.</p>


>*3. Τι γνωρίζετε για τις παραμέτρους α και γ. Τι θα συμβεί αν έχουν τιμές ίσες με 1;*

Υπενθυμίζουμε τη σχέση (1) που περιγράφει την ανανέωση του πίνακα $Q(s,\alpha)$:
<br><br>

$$\tag{2}Q_{\text{new}}(s,\alpha)=Q_{\text{current}}(s,\alpha)+\alpha\cdot \biggl(R(s,\alpha)+\gamma\cdot \max_{\alpha \in A} Q(s',t)-Q(s,\alpha)\biggr)$$.<br><br>

- H παράμετρος $\alpha$ αντιστοιχεί στο βαθμό εκμάθησης και είναι μια τιμή μεταξύ του $(0,1]$. Τιμές του $\alpha$ που βρίσκονται στο $0$ τείνουν να μηδενίσουν τον προσθετέτο στο δεξί μέλος της $(2)$ πράγμα που σημαίνει ότι οι τιμές του πίνακα $Q$ δεν επηρεάζονται σημαντικά από τη νέα πληροφορία που επιφέρουν οι μελλοντικές καταστάσεις. Αντιθέτως, τιμές κοντά στο $1$ επηρέαζουν σημαντικά την ανανέωση του $Q$ ανάλογα με τη νέα πληροφορία που υπεισέρχεται εξαιτίας των επόμενων κινήσεων.

- Η παράμετρος $\gamma$ είναι ο ρυθμός έκπτωσης ο οποίος επηρεάζει την αμοιβή των μελλοντικών ενεργειών του πράκτορα. Ο λόγος ύπαρξης αυτής της παραμέτρου είναι έτσι ώστε ο πράκτορας να επιλέγει τις ενέργεις που επιφέρουν υψηλές επιβραβεύσεις σε λίγα βήματα (ιδανικά στο αμέσως επόμενο βήμα). Τιμές της παραμέτρου κοντά στο $1$ τείνουν τον πράκτορα να επιλέγει όσον το δυνατό συντομότερα τις ενέργειες που του επιφέρουν τις μεγαλύτερες επιβραβεύσεις.


In [17]:
import numpy as np
q_table = np.zeros([env.observation_space.n, env.action_space.n])

In [18]:
%%time
"""Training the agent"""

import random
from IPython.display import clear_output

# Hyperparameters
alpha = 0.1
gamma = 0.6
epsilon = 0.1

# For plotting metrics
all_epochs = []
all_penalties = []

for i in range(1, 100001):
    state = env.reset()

    epochs, penalties, reward, = 0, 0, 0
    done = False
    
    while not done:
        if random.uniform(0, 1) < epsilon:
            action = env.action_space.sample() # Explore action space
        else:
            action = np.argmax(q_table[state]) # Exploit learned values

        next_state, reward, done, info = env.step(action) 
        
        old_value = q_table[state, action]
        next_max = np.max(q_table[next_state])
        
        new_value = (1 - alpha) * old_value + alpha * (reward + gamma * next_max)
        q_table[state, action] = new_value

        if reward == -10:
            penalties += 1

        state = next_state
        epochs += 1
        
    if i % 100 == 0:
        clear_output(wait=True)
        print(f"Episode: {i}")

print("Training finished.\n")

Episode: 100000
Training finished.

CPU times: user 55.8 s, sys: 11.1 s, total: 1min 6s
Wall time: 57.8 s


In [9]:
q_table[328]

array([ -2.40614917,  -2.27325184,  -2.40884107,  -2.36058625,
       -10.432459  , -10.68623986])

Έχοντας εκπαιδεύσει τον πράκτορα εξετάζουμε το πόσο γρήγορα ο πράκτορας φτάνει στην επιθυμητή κατάσταση. Εκτελούμε το πείραμα για 100 επεισόδια και συγκρίνουμε τα αποτελέσματα με τον αλγόριθμο που επιλέγει κινήσεις τυχαία.

In [22]:
"""Evaluate agent's performance after Q-learning"""

total_epochs, total_penalties, total_rewards = 0, 0, 0
episodes = 100

for _ in range(episodes):
    state = env.reset()
    epochs, penalties, reward = 0, 0, 0
    
    done = False
    
    while not done:
        action = np.argmax(q_table[state])
        state, reward, done, info = env.step(action)

        if reward == -10:
            penalties += 1

        epochs += 1
        total_rewards+=reward

    total_penalties += penalties
    total_epochs += epochs

print(f"Results after {episodes} episodes:")
print(f"Average timesteps per episode: {total_epochs / episodes}")
print(f"Average penalties per episode: {total_penalties / episodes}")
print(f"Average reward per move: {total_rewards/total_epochs}")

Results after 100 episodes:
Average timesteps per episode: 12.82
Average penalties per episode: 0.0
Average reward per move: 0.6380655226209049


In [23]:
"""Evaluate agent's performance with random moves"""

total_epochs, total_penalties, total_rewards = 0,0,0
episodes = 100

for _ in range(episodes):
  env.s = 328  # set environment to illustration's state

  epochs = 0
  penalties, reward = 0, 0

  frames = [] # for animation

  done = False

  while not done:
      action = env.action_space.sample()
      state, reward, done, info = env.step(action)

      if reward == -10:
          penalties += 1
      
      # Put each rendered frame into dict for animation
      frames.append({
          'frame': env.render(mode='ansi'),
          'state': state,
          'action': action,
          'reward': reward
          }
      )
      total_rewards+=reward
      epochs+=1
  total_penalties += penalties
  total_epochs += epochs
    
    
print(f"Results after {episodes} episodes:")
print(f"Average timesteps per episode: {total_epochs / episodes}")
print(f"Average penalties per episode: {total_penalties / episodes}")
print(f"Average reward per move: {total_rewards/total_epochs}")

Results after 100 episodes:
Average timesteps per episode: 1818.76
Average penalties per episode: 590.64
Average reward per move: -3.911192240867404


<b><p>4. Συγκρίνετε τους δύο αλγορίθμους με βάση τις παρακάτω μετρικές</p>
<ul>
<li>Μέσος αριθμός παραβάσεων ανά επεισόδιο</li>
<li>Μέσος αριθμός βημάτων ανά διαδρομή</li>
<li>Μέσος αριθμός ανταμοιβών ανά κίνηση</li>
</ul>
<p>Τις παραπάνω συγκρίσεις να τις κάνετε για 100 επεισόδια.</p>
</b>

Στον παρακάτω πίνακα βλέπουμε τα αποτελέσματα των δύο αλγορίθμων, Q-Learning και του αλγορίθμου που επιλέγει με τυχαίο τρόπο όσον αφορά:

1. Τον μέσο αριθμό παραβάσεων ανά επεισόδιο
2. Τον μέσο αριθμό βημάτων ανά διαδρομή και 
3. Τον μέσο αριθμό ανταμοιβών ανά κίνηση.

| **Αλγόριθμος** | **Μέσος αριθμός παραβάσεων** |**Μέσος αριθμός βημάτων**|**Μέσος αριθμός ανταμοιβών ανά κίνηση**|
|:---------------|:-----------------|:-----------------|:-----------------|
| **Q-Learning** |     0       | 12.82 | 0.6380 |
| **Random**      |       590.64     | 1818.76 | -3.9111  |


Όπως βλέπουμε και απ' τον παραπάνω πίνακα η υπεροχή του αλγορίθμου Q-Learning είναι εμφανής ως προς όλες τις μετρικές. Όπως βλέπουμε ο αλγόριθμος Q-Learning κάνει 0 παραβάσεις κατά μέσο όρο και τελειώνει σε περίπου 12 βήματα με μέσο αριθμό ανταμοιβής 0.63. Αντιθέτως, ο αλγόριθμος που επιλέγει με τυχαίο τρόπο βλέπουμε ότι κάνει 590 παραβάσεις κατά μέσο, ολοκληρώνει την αποστολή σε περίπου 1818 βήματα έχοντας -3.9 ζημιά.